README.md
<p align="left"><a href="https://github.com/presseddigital/validateit" target="_blank"><img width="100" height="100" src="resources/img/validateit.svg" alt="Validateit"></a></p>

# Validateit plugin for Craft 3

Supercharged text field validation...

> :mega: **A note from the new (well sort of new) owners Pressed Digital:** At the end of 2019 Fruit Studios Ltd was sold to allow it's founders to shift 100% of their focus to the, now not so side project, findarace.com. As a part of the deal we retained ownership of all Craft plugins and are committed to make them even more awesome. With findarace.com being all Craft, this aligns really nicely, and as a little silver lining we should now finally have some time to launch some of our unreleased plugins!

### Requirements

This plugin requires Craft CMS 3.0.0 or later.

### Installation

To install the plugin, follow these steps:

1.  Install with Composer via:

    composer require presseddigital/validateit

2.  In the Control Panel, go to Settings → Plugins and click the “Install” button for Validateit.

## Validateit Overview

This plugin adds a custom fieldtype which allows you to take control of validation on plain text fields. Validate using the following predefined strings or setup your own custom regex rules. Add custom placeholders and error messages.

1.  Email Address
2.  URL
3.  Phone Number
4.  IP Address
5.  IPv4 Address
6.  IPv6 Address
7.  Facebook Link
8.  Twitter Link
9.  LinkedIn Link
10. Instagram Link
11. Custom Regex Rule

## Configuring Validateit

Once installed, create a new field and choose the Validateit fieldtype. You'll then have the option of configuring how you want to validate the text.

<p align="left"><img width="450px" src="resources/img/configure.png" alt="Configure Validateit"></a></p>

Each field has the option to set a custom placeholder and error message

<p align="left"><img width="450px" src="resources/img/settings.png" alt="Setup Validateit"></a></p>

Custom regex rules allow you to validate any string

<p align="left"><img width="450px" src="resources/img/regex.png" alt="Regex with Validateit"></a></p>

## Using Validateit

Use just as you would any native Plain Text field:

    {{ entry.validateitFieldHandle }}

## Roadmap

If you have any super useful validation rules shout and we can look to add them to the core plugin for everyone to use.

*   [ ] Show default error message and placeholder text in field settings.

Brought to you by [Pressed Digital](https://pressed.digital)


CHANGELOG.md
# Validateit Changelog

## 3.0.0 - 2022-11-03

### Changed
- Bump version numbers to mirror crafts major versions

## 1.0.4 - 2020-11-15

### Changed

*   Transfer ownership to Pressed Digital

## 1.0.3 - 2018-05-18

### Fixed

*   Fixed phone number validation

### Improved

*	Improved url validation

## 1.0.2 - 2018-05-10

### Fixed

*   Fixed our custom validator

## 1.0.1 - 2018-05-01

### Added

*   Docs

## 1.0.0 - 2018-04-30

### Added

*   Initial release


src/Validit.php
<?php
/**
 * Validit plugin for Craft CMS 3.x
 *
 * A super simple field type which allows you toggle existing field types.
 *
 * @link      https://fruitstudios.co.uk
 * @copyright Copyright (c) 2018 Fruit Studios
 */

namespace fruitstudios\validit;

use fruitstudios\validit\fields\ValiditField;

use Craft;
use craft\base\Plugin;
use craft\services\Fields;
use craft\events\RegisterComponentTypesEvent;

use yii\base\Event;

/**
 * Class Validit
 *
 * @author    Fruit Studios
 * @package   Validit
 * @since     1.0.0
 *
 */
class Validit extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * @var Validit
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * @var string
     */
    public $schemaVersion = '1.0.0';

    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        Event::on(
            Fields::class,
            Fields::EVENT_REGISTER_FIELD_TYPES,
            function (RegisterComponentTypesEvent $event) {
                $event->types[] = ValiditField::class;
            }
        );

        Craft::info(
            Craft::t(
                'validit',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

}


src/Validateit.php
<?php
/**
 * Validateit plugin for Craft CMS 3.x
 *
 * A super simple field type which allows you toggle existing field types.
 *
 * @link      https://pressed.digital
 * @copyright Copyright (c) 2018 Pressed Digital
 */

namespace presseddigital\validateit;

use presseddigital\validateit\fields\ValidateitField;

use Craft;
use craft\base\Plugin;
use craft\services\Fields;
use craft\events\RegisterComponentTypesEvent;

use yii\base\Event;

/**
 * Class Validateit
 *
 * @author    Pressed Digital
 * @package   Validateit
 * @since     1.0.0
 *
 */
class Validateit extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * @var Validateit
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * @var string
     */
    public $schemaVersion = '1.0.0';

    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        Event::on(
            Fields::class,
            Fields::EVENT_REGISTER_FIELD_TYPES,
            function (RegisterComponentTypesEvent $event) {
                $event->types[] = ValidateitField::class;
            }
        );

        Craft::info(
            Craft::t(
                'validateit',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

}


src/fields/ValiditField.php
<?php
/**
 * Validit plugin for Craft CMS 3.x
 *
 * A super simple field type which allows you toggle existing field types.
 *
 * @link      https://fruitstudios.co.uk
 * @copyright Copyright (c) 2018 Fruit Studios
 */

namespace fruitstudios\validit\fields;

use fruitstudios\validit\Validit;

use Craft;
use craft\base\ElementInterface;
use craft\base\Field;
use craft\helpers\Db;
use yii\db\Schema;
use craft\helpers\Json;

/**
 * @author    Fruit Studios
 * @package   Validit
 * @since     1.0.0
 */
class ValiditField extends Field
{
    // Public Properties
    // =========================================================================

    public $type;
    public $regex;
    public $message;
    public $placeholder;

    // Static Methods
    // =========================================================================

    public static function displayName(): string
    {
        return Craft::t('validit', 'Validit');
    }

    // Public Methods
    // =========================================================================

    public function rules()
    {
        $rules = parent::rules();
        $rules[] = [['type'], 'required'];
        $rules[] = [['regex'], 'required', 'when' => [$this, 'isCustomType']];
        $rules[] = [['message', 'placeholder', 'regex'], 'string'];
        return $rules;
    }

    public function isCustomType(): bool
    {
        return $this->type == 'custom';
    }

    public function getElementValidationRules(): array
    {
        $message = !empty($this->message) ? $this->message : Craft::t('validit', 'Please provide a valid {type}', [
            'type' => $this->type
        ]);

        switch($this->type)
        {
            case('email'):
                $rule = ['email', 'message' => $message];
                break;
            case('url'):
                $rule = [ValiditUrlValidator::class, 'defaultScheme' => 'http', 'message' => $message];
                break;
            case('tel'):
                $match = '/^(?:\+\d{1,3}|0\d{1,3}|00\d{1,2})?(?:\s?\(\d+\))?(?:[-\/\s.]|\d)+$/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('ip'):
                $rule = ['ip', 'message' => $message];
                break;
            case('ipv4'):
                $rule = ['ip', 'ipv6' => false, 'message' => $message];
                break;
            case('ipv6'):
                $rule = ['ip', 'ipv4' => false, 'message' => $message];
                break;
            case('facebook'):
                $match = '/(?:(?:http|https):\/\/)?(?:www.)?facebook.com\/(?:(?:\w)*#!\/)?(?:pages\/)?(?:[?\w\-]*\/)?(?:profile.php\?id=(?=\d.*))?([\w\-]*)?/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('twitter'):
                $match = '/^http(?:s)?:\/\/(?:www\.)?twitter\.com\/([a-zA-Z0-9_]+)/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('instagram'):
                $match = '/(?:(?:http|https):\/\/)?(?:www.)?(?:instagram.com|instagr.am)\/([A-Za-z0-9-_]+)/i';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('linkedin'):
                $match = '/^http(?:s)?:\/\/[a-z]{2,3}\\.linkedin\\.com\\/.*$/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('custom'):
                $rule = ['match', 'pattern' => $this->regex, 'message' => $message];
                break;
            default:
                $rule = null;
                break;
        }

        return [$rule];
    }

    public function getContentColumnType(): string
    {
        return Schema::TYPE_TEXT;
    }

    public function normalizeValue($value, ElementInterface $element = null)
    {
        return $value;
    }

    public function serializeValue($value, ElementInterface $element = null)
    {
        return parent::serializeValue($value, $element);
    }

    /**
     * @inheritdoc
     */
    public function getSettingsHtml()
    {
        // Render the settings template
        return Craft::$app->getView()->renderTemplate(
            'validit/_settings',
            [
                'field' => $this,
            ]
        );
    }

    /**
     * @inheritdoc
     */
    public function getInputHtml($value, ElementInterface $element = null): string
    {
        // Get our id and namespace
        $id = Craft::$app->getView()->formatInputId($this->handle);

        // Render the input template
        return Craft::$app->getView()->renderTemplate(
            'validit/_input',
            [
                'id' => $id,
                'name' => $this->handle,
                'value' => $value,
                'field' => $this,
            ]
        );
    }

    public function getTypes()
    {

        return [
            'email' => [
                'label' => Craft::t('validit', 'Email Address'),
                'placeholder' => Craft::t('validit', 'email@domain.com'),
                'handle' => 'email',
            ],
            'url' => [
                'label' => Craft::t('validit', 'URL'),
                'placeholder' => Craft::t('validit', 'https://domain.com'),
                'handle' => 'url',
            ],
            'telephone' => [
                'label' => Craft::t('validit', 'Telephone'),
                'placeholder' => Craft::t('validit', '+44(0)0000 000000'),
                'handle' => 'telephone',
            ],
            'ip' => [
                'label' => Craft::t('validit', 'IP Address (Any)'),
                'placeholder' => Craft::t('validit', '192.168.0.1, 2001:0db8:85a3:0000:0000:8a2e:0370:7334'),
                'handle' => 'ip',
            ],
            'ipv4' => [
                'label' => Craft::t('validit', 'IP Address (IPV4)'),
                'placeholder' => Craft::t('validit', '192.168.0.1'),
                'handle' => 'ipv4',
            ],
            'ipv6' => [
                'label' => Craft::t('validit', 'IP Address (IPV6)'),
                'placeholder' => Craft::t('validit', '2001:0db8:85a3:0000:0000:8a2e:0370:7334'),
                'handle' => 'ipv6',
            ],
            'facebook' => [
                'label' => Craft::t('validit', 'Facebook Url'),
                'placeholder' => Craft::t('validit', 'https://www.facebook.com/username'),
                'handle' => 'facebook',
            ],
            'twitter' => [
                'label' => Craft::t('validit', 'Twitter Url'),
                'placeholder' => Craft::t('validit', 'https://twitter.com/username'),
                'handle' => 'twitter',
            ],
            'instagram' => [
                'label' => Craft::t('validit', 'Instagram Url'),
                'placeholder' => Craft::t('validit', 'https://www.instagram.com/username'),
                'handle' => 'instagram',
            ],
            'linkedin' => [
                'label' => Craft::t('validit', 'LinkedIn Url'),
                'placeholder' => Craft::t('validit', 'https://www.linkedin.com/in/username'),
                'handle' => 'linkedin',
            ],
            'custom' => [
                'label' => Craft::t('validit', 'Custom Regex'),
                'placeholder' => Craft::t('validit', $this->name),
                'handle' => 'custom',
            ]
        ];
    }

    public function getTypeOptions()
    {
        $options = [];
        foreach ($this->getTypes() as $type => $value)
        {
            $options[] = [
                'label' => $type->label,
                'value' => $type->handle,
            ];
        }
        return $options;
    }

}


src/validators/UrlValidator.php
<?php
namespace presseddigital\validateit\validators;

use Craft;
use yii\validators\UrlValidator as YiiUrlValidator;

class UrlValidator extends YiiUrlValidator
{
    // Public Methods
    // =========================================================================

    public function __construct(array $config = [])
    {

        // Override the $pattern regex so that a TLD is not required, and the protocol may be relative and can be a #
        if (!isset($config['pattern'])) {
            $config['pattern'] = '/^(\/|#|(?:(?:{schemes}:)?\/\/(([A-Z0-9][A-Z0-9_-]*)(\.[A-Z0-9][A-Z0-9_-]*)+)?|\/)[^\s]*$)/i';
        }

        // Enable support for validating international domain names if the intl extension is available.
        if (!isset($config['enableIDN']) && Craft::$app->getI18n()->getIsIntlLoaded() && defined('INTL_IDNA_VARIANT_UTS46')) {
            $config['enableIDN'] = true;
        }

        parent::__construct($config);
    }

    public function validateValue($value)
    {
        if($value === '#' || $value === '/') {
            $this->defaultScheme = null;
            return null;
        }

        // Add support for protocol-relative URLs, # or /
        if ($this->defaultScheme !== null && strpos($value, '/') === 0) {
            $this->defaultScheme = null;
        } else {
            $this->pattern = '/^{schemes}:\/\/(([A-Z0-9][A-Z0-9_-]*)(\.[A-Z0-9][A-Z0-9_-]*)+)(?::\d{1,5})?(?:$|[?\/#])/i';
        }

        return parent::validateValue($value);
    }

}


src/fields/ValidateitField.php
<?php
/**
 * Validateit plugin for Craft CMS 3.x
 *
 * A super simple field type which allows you toggle existing field types.
 *
 * @link      https://fruitstudios.co.uk
 * @copyright Copyright (c) 2018 Fruit Studios
 */

namespace fruitstudios\validateit\fields;

use fruitstudios\validateit\Validateit;
use fruitstudios\validateit\validators\UrlValidator as ValidateitUrlValidator;

use Craft;
use craft\base\ElementInterface;
use craft\base\PreviewableFieldInterface;
use craft\base\SortableFieldInterface;
use craft\base\Field;
use craft\helpers\Db;
use craft\helpers\Json;

use yii\db\Schema;

/**
 * @author    Fruit Studios
 * @package   Validateit
 * @since     1.0.0
 */
class ValidateitField extends Field implements PreviewableFieldInterface, SortableFieldInterface
{
    // Public Properties
    // =========================================================================

    public $type;
    public $regex;
    public $message;
    public $placeholder;

    // Static Methods
    // =========================================================================

    public static function displayName(): string
    {
        return Craft::t('validateit', 'Validateit');
    }

    // Public Methods
    // =========================================================================

    public function rules()
    {
        $rules = parent::rules();
        $rules[] = [['type'], 'required'];
        $rules[] = [['regex'], 'required', 'when' => [$this, 'isCustomType']];
        $rules[] = [['message', 'placeholder', 'regex'], 'string'];
        return $rules;
    }

    public function isCustomType(): bool
    {
        return $this->type == 'custom';
    }

    public function getTypeArray(): array
    {
        return $this->getTypes()[$this->type] ?? null;
    }

    public function getElementValidationRules(): array
    {
        $message = !empty($this->message) ? $this->message : ( $this->getTypeArray()['error'] ?? null );

        switch($this->type)
        {
            case('email'):
                $rule = ['email', 'message' => $message];
                break;
            case('url'):
                $rule = [ValidateitUrlValidator::class, 'defaultScheme' => 'http', 'message' => $message];
                break;
            case('phone'):
                $match = '/^(?:\+\d{1,3}|0\d{1,3}|00\d{1,2})?(?:\s?\(\d+\))?(?:[-\/\s.]|\d)+$/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('ip'):
                $rule = ['ip', 'message' => $message];
                break;
            case('ipv4'):
                $rule = ['ip', 'ipv6' => false, 'message' => $message];
                break;
            case('ipv6'):
                $rule = ['ip', 'ipv4' => false, 'message' => $message];
                break;
            case('facebook'):
                $match = '/(?:(?:http|https):\/\/)?(?:www.)?facebook.com\/(?:(?:\w)*#!\/)?(?:pages\/)?(?:[?\w\-]*\/)?(?:profile.php\?id=(?=\d.*))?([\w\-]*)?/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('twitter'):
                $match = '/^http(?:s)?:\/\/(?:www\.)?twitter\.com\/([a-zA-Z0-9_]+)/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('instagram'):
                $match = '/(?:(?:http|https):\/\/)?(?:www.)?(?:instagram.com|instagr.am)\/([A-Za-z0-9-_]+)/i';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('linkedin'):
                $match = '/^http(?:s)?:\/\/[a-z]{2,3}\\.linkedin\\.com\\/.*$/';
                $rule = ['match', 'pattern' => $match, 'message' => $message];
                break;
            case('custom'):
                $rule = ['match', 'pattern' => $this->regex, 'message' => $message];
                break;
            default:
                $rule = null;
                break;
        }

        return [$rule];
    }

    public function getContentColumnType(): string
    {
        return Schema::TYPE_TEXT;
    }

    public function normalizeValue($value, ElementInterface $element = null)
    {
        return $value;
    }

    public function serializeValue($value, ElementInterface $element = null)
    {
        return parent::serializeValue($value, $element);
    }

    /**
     * @inheritdoc
     */
    public function getSettingsHtml()
    {
        // Render the settings template
        return Craft::$app->getView()->renderTemplate(
            'validateit/_settings',
            [
                'field' => $this,
            ]
        );
    }

    /**
     * @inheritdoc
     */
    public function getInputHtml($value, ElementInterface $element = null): string
    {
        // Get our id and namespace
        $id = Craft::$app->getView()->formatInputId($this->handle);

        // Render the input template
        return Craft::$app->getView()->renderTemplate(
            'validateit/_input',
            [
                'id' => $id,
                'name' => $this->handle,
                'value' => $value,
                'field' => $this,
            ]
        );
    }

    public function getTypes()
    {
        return [
            'email' => [
                'label' => Craft::t('validateit', 'Email Address'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'email address')
                ]),
                'placeholder' => Craft::t('validateit', 'email@domain.com'),
                'handle' => 'email',
            ],
            'url' => [
                'label' => Craft::t('validateit', 'URL'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'link')
                ]),
                'placeholder' => Craft::t('validateit', 'https://domain.com'),
                'handle' => 'url',
            ],
            'phone' => [
                'label' => Craft::t('validateit', 'Phone Number'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'phone number')
                ]),
                'placeholder' => Craft::t('validateit', '+44(0)0000 000000'),
                'handle' => 'phone',
            ],
            'ip' => [
                'label' => Craft::t('validateit', 'IP Address'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'IP address')
                ]),
                'placeholder' => Craft::t('validateit', '192.168.0.1, 2001:0db8:85a3:0000:0000:8a2e:0370:7334'),
                'handle' => 'ip',
            ],
            'ipv4' => [
                'label' => Craft::t('validateit', 'IP Address (IPv4)'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'IPv4 address')
                ]),
                'placeholder' => Craft::t('validateit', '192.168.0.1'),
                'handle' => 'ipv4',
            ],
            'ipv6' => [
                'label' => Craft::t('validateit', 'IP Address (IPv6)'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'IPv6 address')
                ]),
                'placeholder' => Craft::t('validateit', '2001:0db8:85a3:0000:0000:8a2e:0370:7334'),
                'handle' => 'ipv6',
            ],
            'facebook' => [
                'label' => Craft::t('validateit', 'Facebook Url'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'Facebook link')
                ]),
                'placeholder' => Craft::t('validateit', 'https://www.facebook.com/username'),
                'handle' => 'facebook',
            ],
            'twitter' => [
                'label' => Craft::t('validateit', 'Twitter Url'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'Twitter link')
                ]),
                'placeholder' => Craft::t('validateit', 'https://twitter.com/username'),
                'handle' => 'twitter',
            ],
            'instagram' => [
                'label' => Craft::t('validateit', 'Instagram Url'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'Instagram link')
                ]),
                'placeholder' => Craft::t('validateit', 'https://www.instagram.com/username'),
                'handle' => 'instagram',
            ],
            'linkedin' => [
                'label' => Craft::t('validateit', 'LinkedIn Url'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'LinkedIn link')
                ]),
                'placeholder' => Craft::t('validateit', 'https://www.linkedin.com/in/username'),
                'handle' => 'linkedin',
            ],
            'custom' => [
                'label' => Craft::t('validateit', 'Custom Regex'),
                'error' => Craft::t('validateit', 'Please provide a valid {type}.', [
                    'type' => Craft::t('validateit', 'value')
                ]),
                'placeholder' => Craft::t('validateit', $this->name),
                'handle' => 'custom',
            ]
        ];
    }

    public function getTypeOptions()
    {
        $options = [];
        foreach ($this->getTypes() as $type => $value)
        {
            $options[] = [
                'label' => $type->label,
                'value' => $type->handle,
            ];
        }
        return $options;
    }

}
